plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,296 @@
+    {
+        Map<String, String> icebergTableProperties = new HashMap<>();
+
+        icebergTableProperties.putAll(hiveTableProperties);
+        EXCLUDED_PROPERTIES.forEach(icebergTableProperties::remove);
+
+        icebergTableProperties.put("provider", "iceberg");
+        icebergTableProperties.put("migrated", "true");
+        icebergTableProperties.putIfAbsent("location", location);
+        icebergTableProperties.put(DEFAULT_NAME_MAPPING, toJson(nameMapping));
+        icebergTableProperties.put(FORMAT_VERSION, String.valueOf(FORMAT_VERSION_SUPPORT_MAX));
+
+        return ImmutableMap.copyOf(icebergTableProperties);
+    }
+
+    private Schema toIcebergSchema(List<Column> columns)
+    {
+        List<Types.NestedField> icebergColumns = new ArrayList<>();
+        for (Column column : columns) {
+            int index = icebergColumns.size();
+            org.apache.iceberg.types.Type type = toIcebergType(typeManager.getType(column.getType().getTypeSignature()));
+            Types.NestedField field = Types.NestedField.of(index, false, column.getName(), type, column.getComment().orElse(null));
+            icebergColumns.add(field);
+        }
+        org.apache.iceberg.types.Type icebergSchema = Types.StructType.of(icebergColumns);
+        AtomicInteger nextFieldId = new AtomicInteger(1);
+        icebergSchema = TypeUtil.assignFreshIds(icebergSchema, nextFieldId::getAndIncrement);
+        return new Schema(icebergSchema.asStructType().fields());
+    }
+
+    public Map<String, Optional<Partition>> listAllPartitions(HiveMetastore metastore, io.trino.plugin.hive.metastore.Table table)
+    {
+        List<String> partitionNames = table.getPartitionColumns().stream().map(Column::getName).collect(toImmutableList());
+        Optional<List<String>> partitions = metastore.getPartitionNamesByFilter(table.getDatabaseName(), table.getTableName(), partitionNames, TupleDomain.all());
+        if (partitions.isEmpty()) {
+            return ImmutableMap.of();
+        }
+        return metastore.getPartitionsByNames(table, partitions.get());
+    }
+
+    private List<DataFile> buildDataFiles(ConnectorSession session, HiveStorageFormat format, String location, PartitionSpec partitionSpec, String partitionKey)
+            throws IOException
+    {
+        Path locationPath = new Path(location);
+        FileSystem fileSystem = hdfsEnvironment.getFileSystem(new HdfsContext(session), locationPath);
+        List<FileStatus> fileStatus = Arrays.stream(fileSystem.listStatus(locationPath, HIDDEN_PATH_FILTER))
+                .filter(FileStatus::isFile)
+                .collect(toImmutableList());
+        DataFile[] datafiles = new DataFile[fileStatus.size()];
+
+        Tasks.range(fileStatus.size())
+                .stopOnFailure()
+                .throwFailureWhenFinished()
+                .run(index -> datafiles[index] = buildDataFile(fileStatus.get(index), partitionKey, partitionSpec, format.name()));
```

Comment:
Does this introduce parallelism? I'm thinking for large tables this is going to be pretty slow. 

It might be overkill, but the HiveSplitLoader has a lot of logic to parallelize listing files across a few threads. Maybe we could reuse that here?

Reply:
No, it doesn't. I had the same concern, but avoided the implementation to start with small code. Let me confirm again.